File: benchmarks/correctness/r/diversity.R
# Shannon and Simpson diversity over GFF feature types. Output JSON.
library(jsonlite)

lines <- readLines("data/annotations.gff3")
lines <- lines[!startsWith(lines, "#") & nzchar(trimws(lines))]
types <- sapply(strsplit(lines, "\t", fixed = TRUE), function(p) if (length(p) >= 3) p[3] else NA)
types <- types[!is.na(types)]

counts <- table(types)
total <- sum(counts)
ps <- as.numeric(counts[order(names(counts))]) / total
shannon <- -sum(ps * log(ps))
simpson <- 1 - sum(ps^2)

cat(toJSON(list(n_categories = length(ps), n_observations = as.integer(total),
                shannon = round(shannon, 9), simpson = round(simpson, 9)),
           auto_unbox = TRUE, pretty = TRUE, digits = 10))
cat("\n")

File: benchmarks/correctness/python/diversity.py
"""Shannon and Simpson diversity over GFF feature types. Output JSON."""
import json
import sys
import math
from collections import Counter

counts = Counter()
total = 0
with open("data/annotations.gff3") as f:
    for line in f:
        if line.startswith("#") or not line.strip():
            continue
        parts = line.rstrip("\n").split("\t")
        if len(parts) >= 3:
            counts[parts[2]] += 1
            total += 1

ps = [counts[k] / total for k in sorted(counts)]
shannon = -sum(p * math.log(p) for p in ps)
simpson = 1 - sum(p * p for p in ps)

json.dump({"n_categories": len(ps), "n_observations": total,
           "shannon": round(shannon, 9), "simpson": round(simpson, 9)},
          sys.stdout, indent=2)
print()
